Add error_inputs_sparse method to OpInfo

pearu opened the pull request:
Per https://github.com/pytorch/pytorch/pull/98288#discussion_r1170553576, this PR introduces `OpInfo.error_inputs_sparse_func` attribute for sparse inputs in parallel to the `OpInfo.error_inputs_func` attribute which is used for strided inputs.

These attributes are kept separate as the existing testing framework that calls `error_inputs_func` may apply operations to inputs that are unsupported for sparse tensors (e.g. as in test/functorch/).


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98292
* #100394
* #100393
* #100392
* #100391
* __->__ #100389



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer

pearu committed:
Add error_inputs_sparse method to OpInfo

[ghstack-poisoned]

cpuhrsch commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -1168,6 +1171,18 @@ def error_inputs(self, device, **kwargs):
+
+    def supports_sparse_layout(self, layout):
+        """Return True if OpInfo supports the specified sparse layout."""
+        layout_name = str(layout).split(".")[-1].replace("_coo", "")
```
Why does this replace "_coo"?

pearu replied:
For historical reasons, OpInfo has the attribute `supports_sparse` which really means "supports sparse COO tensors". On the other hand, we have layout `torch.sparse_coo`. So, the `replace "_coo"` will relate the `OpInfo.supports_sparse` and `torch.sparse_coo`.

I think we should replace `OpInfo.supports_sparse` with `OpInfo.supports_sparse_coo` for clarity ("sparse" has too many different meanings, e.g. non-dense or COO) but this is slightly OT for this PR.

pmeier replied:
Nit: I would put the `replace` down below to the `getattr` call, since, as you explained, this is only related to the intricacies of this class rather than layouts in general. Plus, putting your explanation in a code comment would also be nice.

pearu replied:
Applied the nit.

cpuhrsch commented:
@pytorchbot merge

pmeier commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -1168,6 +1171,18 @@ def error_inputs(self, device, **kwargs):
+    def supports_sparse_layout(self, layout):
+        """Return True if OpInfo supports the specified sparse layout."""
+        layout_name = str(layout).split(".")[-1].replace("_coo", "")
+        return getattr(self, "supports_" + layout_name)
```
Nit
```suggestion
        return getattr(self, f"supports_{layout_name}")
```

pearu replied:
Fixed.

pearu committed:
Update on "Add error_inputs_sparse method to OpInfo"




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge

pearu committed:
Update on "Add error_inputs_sparse method to OpInfo"

Per https://github.com/pytorch/pytorch/pull/98288#discussion_r1170553576, this PR introduces `OpInfo.error_inputs_sparse_func` attribute for sparse inputs in parallel to the `OpInfo.error_inputs_func` attribute which is used for strided inputs.

These attributes are kept separate as the existing testing framework that calls `error_inputs_func` may apply operations to inputs that are unsupported for sparse tensors (e.g. as in test/functorch/).





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer

[ghstack-poisoned]

pearu commented:
@pytorchbot merge
